pkg/resourcegenerator/core/pod.go

```diff
@@ -44,16 +45,20 @@ func CreatePodSpec(container corev1.Container, volumes []corev1.Volume, serviceA
 		ImagePullSecrets:  []corev1.LocalObjectReference{{Name: "github-auth"}},
 		SchedulerName:     corev1.DefaultSchedulerName,
 		PriorityClassName: fmt.Sprintf("skip-%s", priority),
-		Affinity: &corev1.Affinity{
-			PodAntiAffinity: &corev1.PodAntiAffinity{
-				PreferredDuringSchedulingIgnoredDuringExecution: []corev1.WeightedPodAffinityTerm{
-					affinityTermForAppAndKey(container.Name, Hostname),
-					affinityTermForAppAndKey(container.Name, OnPremFailureDomain),
-				},
-			},
-		},
 	}
 
+	// Global feature flag
+	if !flags.FeatureFlags.DisablePodTopologySpreadConstraints {
+		// Allow override per application
+		if !podSettings.DisablePodSpreadTopologyConstraints {
+			p.TopologySpreadConstraints = []corev1.TopologySpreadConstraint{
+				spreadConstraintForAppAndKey(container.Name, Hostname),
+				spreadConstraintForAppAndKey(container.Name, OnPremFailureDomain),
```

Comment:
When does the failure-domain stuff kick in? I what scenarios will there be multiple?

Reply:
> You can use topology spread constraints to control how Pods are spread across your cluster among failure-domains such as regions, zones, nodes, and other user-defined topology domains. This can help to achieve high availability as well as efficient resource utilization.
> – [K8s doc](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/)

As far as I know, these constraints are are only applied when scheduling pods onto nodes (in addition to CPU/mem requests, priority classes, persistent storage etc.). 

As for multiple failure domains:
- `kubernetes.io/hostname`: We want to spread pods over more nodes in order to spread resource usage and guard against individual node failures (VM down, flaky network etc). This is a K8s standard label.
- `onprem.gke.io/failure-domain-name`: This is an Anthos/GKE On-Prem specific label. In addition to individual VM failures we want to maximize the pod spread over individual physical ESXi hosts (containing one or more VMs acting as K8s nodes).
- [my understanding] If one of the labels is not present on the node it's simply not taken into account, resulting in using only the other specified node labels (in this code, the hostname label). This should be verified by running a build in one of the cloud clusters and see pods be scheduled over the various available hosts. If this understanding is correct, a new PR should be issued to also take the K8s standard label `topology.kubernetes.io/zone` (populated by either the Kubelet or the cloud-controller-manager provided by GKE). It's not implemented today because it's missing in GKE On-Prem.

Reply:
> onprem.gke.io/failure-domain-name: This is an Anthos/GKE On-Prem specific label. In addition to individual VM failures we want to maximize the pod spread over individual physical ESXi hosts (containing one or more VMs acting as K8s nodes).

And the ESXi hosts have different values for this label? Sounds like a good idea!

Reply:
Yes, they have :-) It's explicitly mentioned in the [documentation](https://cloud.google.com/anthos/clusters/docs/on-prem/latest/concepts/high-availability-disaster-recovery#user_cluster) and can be verified by running this command against an environment, e.g. `prod` or `dev`:
```shell
kubectl get nodes -o json | jq '.items | map({name: .metadata.name, failureDomain: .metadata.labels["onprem.gke.io/failure-domain-name"]}) | group_by(.failureDomain) | map({failureDomain: .[0].failureDomain, nodes: map(.name)})'
```